File: src/usage/session/contracts.ts
import type { AgentSessionRef } from "../../agent/contracts";

export interface SessionStatisticsQuery { session: AgentSessionRef; transcriptPath: string | null }
export interface SessionTokens { input: number | null; cachedInput: number | null; cacheWrite: number | null; output: number | null }
export interface SessionStatistics {
  session: AgentSessionRef;
  tokens: SessionTokens;
  models: string[];
  currentModel: string | null;
  tools: { name: string; calls: number }[];
  toolCount: number | null;
  updatedAt: number | null;
  observedAt: number;
  sourceFiles: number;
  scannedBytes: number;
  pending: boolean;
  skippedLines: number;
  note: string | null;
}

export interface StatisticsView { report: SessionStatistics | null; loading: boolean; error: string | null }
export const EMPTY_STATISTICS: StatisticsView = { report: null, loading: false, error: null };

File: src-tauri/src/usage/session/contracts.rs
use crate::agent::AgentSessionRef;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionStatisticsQuery {
    pub session: AgentSessionRef,
    pub transcript_path: Option<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionTokens {
    pub input: Option<u64>,
    pub cached_input: Option<u64>,
    pub cache_write: Option<u64>,
    pub output: Option<u64>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ToolStatistics {
    pub name: String,
    pub calls: u64,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionStatistics {
    pub session: AgentSessionRef,
    pub tokens: SessionTokens,
    pub models: Vec<String>,
    pub current_model: Option<String>,
    pub tools: Vec<ToolStatistics>,
    pub tool_count: Option<u64>,
    pub updated_at: Option<i64>,
    pub observed_at: i64,
    pub source_files: usize,
    pub scanned_bytes: u64,
    pub pending: bool,
    pub skipped_lines: usize,
    pub note: Option<String>,
}
